### PR TITLE
[Add] Refactor evaluation for learning-based methods

### PR DESCRIPTION
**Pull request by wqyin**
1. Refactor evaluation for learning-based methods
2. Fix file saving for predicted keypoints3d during the evaluation with multiple GPUs

### CODE REVIEW

File: xrmocap/core/evaluation/mvp_evaluation.py [hunk 1]
```diff
@@ -54,14 +277,12 @@ def evaluate_map(self, pred_kps3d: torch.Tensor) \
...
-                                                                       3] >= 0]
+            pred_frame_kps3d_valid = pred_kps3d[frame_idx_all_scenes].copy()
 
             for pred_person_kps3d in pred_frame_kps3d_valid:
                 mpjpes = []
                 for gt_person_kps3d in gt_frame_kps3d:
-                    vis = gt_person_kps3d[:, -1] > self.kps_thr
+                    vis = gt_person_kps3d[:, -1] > threshold
 
                     mpjpe = np.mean(
```
Comment (jiaqiAA) [inline xrmocap/core/evaluation/mvp_evaluation.py:287]: Can we calc mpjpe or others from [metrics.py](https://github.com/openxrlab/xrmocap/blob/main/xrmocap/core/evaluation/metrics.py)?

Comment (wqyin) [reply xrmocap/core/evaluation/mvp_evaluation.py:341]: Replaced the PCP and MPJPE calculation for Shelf and Campus with metrics.py functions in [8696023](https://github.com/openxrlab/xrmocap/pull/47/commits/8696023dc3de99fcd178d41f67a14d8b7250cd8c).

File: xrmocap/core/evaluation/mvp_evaluation.py [hunk 2]
```diff
@@ -1,67 +1,341 @@
...
+            if not is_train:
+                pred_file = os.path.join(self.output_dir, 'pred_kps3d.npz')
+                gt_file = os.path.join(self.output_dir, 'gt_kps3d.npz')
+
+                self.logger.info(f'Saving 3D keypoints to: {pred_file}')
+                keypoints3d_pred.dump(pred_file)
+                self.logger.info(f'Saving 3D keypoints to: {gt_file}')
+                keypoints3d_gt.dump(gt_file)
+
+        # quantitative evaluation and print result
```
Comment (yl-1993) [inline xrmocap/core/evaluation/mvp_evaluation.py:164]: Add indent

**Review by yl-1993 — APPROVED** [review]

File: xrmocap/core/evaluation/mvp_evaluation.py [hunk 3]
```diff
@@ -1,67 +1,341 @@
...
 
-from xrmocap.data.dataset.base_dataset import BaseDataset
+from xrmocap.core.evaluation.align_keypoints3d import align_keypoints3d
+from xrmocap.data_structure.keypoints import Keypoints
+from xrmocap.model.architecture.base_architecture import BaseArchitecture
+from xrmocap.utils.distribute_utils import collect_results, is_main_process
+from xrmocap.utils.mvp_utils import (
+    AverageMeter, convert_result_to_kps, norm2absolute,
+)
+from .metrics import calc_limbs_accuracy, evaluate
```
Comment (yl-1993) [inline xrmocap/core/evaluation/mvp_evaluation.py:19]: The name `evaluate` is a little general, which contains mpjpe, pampjpe and pck at the same time. Any better name to describe it?

File: xrmocap/core/evaluation/mvp_evaluation.py (hunk 2 above)
Comment (wqyin) [reply xrmocap/core/evaluation/mvp_evaluation.py]: Fixed in [d58a15a](https://github.com/openxrlab/xrmocap/pull/47/commits/d58a15a9e9978d496b1e4abb214c07c375049e06).

File: xrmocap/core/evaluation/mvp_evaluation.py (hunk 3 above)
Comment (yl-1993) [reply xrmocap/core/evaluation/mvp_evaluation.py:19]: After offline discussion, it will be left to #50

**Pull request merged**